### ISSUE TITLE
training on two different inputs gives some bad results

### ISSUE THREAD

**Issue by riyaj8888** [keras-team/keras#13066]
`import numpy as np
import os
import random
import cv2
from keras import backend as K
from keras.layers import Layer ,core
from imutils import paths
from keras.preprocessing.image import ImageDataGenerator
from keras.optimizers import Adam
from sklearn.model_selection import train_test_split
from keras.preprocessing.image import img_to_array
from keras.utils import to_categorical
from keras import layers
from keras.preprocessing.image import ImageDataGenerator
from keras.layers import Input,  Concatenate,Reshape,Add, Dense, UpSampling2D,Dropout,Activation, ZeroPadding2D, BatchNormalization, Flatten, Conv2D, AveragePooling1D,AveragePooling2D, MaxPooling2D, GlobalMaxPooling2D
from keras.models import Model, load_model
from keras.preprocessing import image
from keras.utils import layer_utils
from keras.utils.data_utils import get_file
from keras.applications.imagenet_utils import preprocess_input
import pydot
from IPython.display import SVG
from keras.utils.vis_utils import model_to_dot
from keras.utils import plot_model
from keras.initializers import Constant , Orthogonal
from keras.layers import LeakyReLU
from keras.callbacks import Callback , ReduceLROnPlateau , TensorBoard , ModelCheckpoint
from keras import regularizers
from keras import optimizers
from sklearn.model_selection import train_test_split
from keras.preprocessing.image import img_to_array
#######################################################################################################
from sklearn.utils import resample
from imutils import paths
from keras import initializers

class MaxoutConv2D(Layer):
    """
    Convolution Layer followed by Maxout activation as described
    in https://arxiv.org/abs/1505.03540.
   
    Parameters
    ----------
   
    kernel_size: kernel_size parameter for Conv2D
    output_dim: final number of filters after Maxout
    nb_features: number of filter maps to take the Maxout over; default=4
    padding: 'same' or 'valid'
    first_layer: True if x is the input_tensor
    input_shape: Required if first_layer=True
   
    """
   
    def __init__(self, kernel_size, output_dim, nb_features=4, padding='valid', **kwargs):
       
        self.kernel_size = kernel_size
        self.output_dim = output_dim
        self.nb_features = nb_features
        self.padding = padding
        super(MaxoutConv2D, self).__init__(**kwargs)

    def call(self, x):

        output = None
        for _ in range(self.output_dim):
           
            conv_out = Conv2D(self.nb_features, self.kernel_size, padding=self.padding)(x)
            maxout_out = K.max(conv_out, axis=-1, keepdims=True)

            if output is not None:
                output = K.concatenate([output, maxout_out], axis=-1)

            else:
                output = maxout_out

        return output

    def compute_output_shape(self, input_shape):
        input_height= input_shape[1]
        input_width = input_shape[2]

        if(self.padding == 'same'):
            output_height = input_height
            output_width = input_width
        elif(input_height == None or input_width == None):
            return (input_shape[0], None, None, self.output_dim)

        else:
            output_height = input_height - self.kernel_size[0] + 1
            output_width = input_width - self.kernel_size[1] + 1

        return (input_shape[0], output_height, output_width, self.output_dim)

############################################################

(img_height ,img_width )= (784,784)
# feature extraction from gray scale image
inputs = Input(shape = (img_height ,img_width,3))

conv1 = Conv2D(16, (3,3), activation = 'relu', padding = "SAME")(inputs)
pool1 = MaxPooling2D(pool_size = (2,2), strides = 2)(conv1)
conv2 = Conv2D(32, (3,3), activation = 'relu', padding = "SAME")(pool1)
pool2 = MaxPooling2D(pool_size = (5,5), strides = 2)(conv2)
conv3 = Conv2D(64, (3,3), activation = 'relu', padding = "SAME")(pool2)
conv3=MaxoutConv2D((3,3),16)(conv3)
pool3 = MaxPooling2D(pool_size = (5,5), strides = 2)(conv3)
conv4= Conv2D(32, (3,3), activation = 'relu', padding = "SAME")(pool3)
conv4=MaxoutConv2D((3,3),16)(conv4)
pool4 = MaxPooling2D(pool_size = (5,5), strides = 2)(conv4)

flat_1 = Flatten()(pool4)

# feature extraction from RGB image
inputs_2 = Input(shape = (img_height ,img_width,3))

conv1_2 = Conv2D(16, (3,3), activation = 'relu', padding = "SAME")(inputs_2)
pool1_2 = MaxPooling2D(pool_size = (2,2), strides = 2)(conv1_2)
conv2_2 = Conv2D(32, (3,3), activation = 'relu', padding = "SAME")(pool1_2)
pool2_2 = MaxPooling2D(pool_size = (5,5), strides = 2)(conv2_2)
conv2_3 = Conv2D(64, (3,3), activation = 'relu', padding = "SAME")(pool2_2)
conv2_3=MaxoutConv2D((3,3),16)(conv2_3)
pool3_2 = MaxPooling2D(pool_size = (5,5), strides = 2)(conv2_3)

conv2_4= Conv2D(32, (3,3), activation = 'relu', padding = "SAME")(pool3_2)
conv2_4=MaxoutConv2D((3,3),16)(conv2_4)
pool3_3 = MaxPooling2D(pool_size = (5,5), strides = 2)(conv2_4)

flat_2 = Flatten()(pool3_3)

# concatenate both feature layers and define output layer after some dense layers
concat = Add()([flat_1,flat_2])
dense1 = Dense(512, activation = 'relu')(concat)
dense2 = Dense(128, activation = 'relu')(dense1)
dense3 = Dense(32, activation = 'relu')(dense2)
output = Dense(2, activation = 'softmax')(dense3)

# create model with two inputs
model = Model([inputs,inputs_2], output)
model.summary()
input_imgen = ImageDataGenerator(rescale = 1./255, )
                                 #shear_range = 0.2, 
                                 #zoom_range = 0.2,
                                 #rotation_range=5.,
                                 #horizontal_flip = True)

test_imgen = ImageDataGenerator(rescale = 1./255)

def generate_generator_multiple(generator,dir1, dir2, batch_size, img_height,img_width):

  genX1 = generator.flow_from_directory(dir1,
                                        target_size = (img_height,img_width),
                                        class_mode = 'categorical',
                                        batch_size = batch_size,
                                        shuffle=False, 
                                        seed=7)

  genX2 = generator.flow_from_directory(dir2,
                                        target_size = (img_height,img_width),
                                        class_mode = 'categorical',
                                        batch_size = batch_size,
                                        shuffle=False, 
                                        seed=7)
  while True:
          X1i = genX1.next()
          X2i = genX2.next()
          # print(X2i)
          yield [X1i[0], X2i[0]], X2i[1]  #Yield both images and their mutual label
          
# train_dir_1 ='cats_dogs/kagglecatsanddogs_3367a/PetImages/data_cat_dogs_100/train_100/'
# train_dir_2 ='cats_dogs/kagglecatsanddogs_3367a/PetImages/data_cat_dogs_100/train_gray_100/'
          

inputgenerator=generate_generator_multiple(generator=input_imgen,
                                         dir1=train_dir_1,
                                         dir2=train_dir_2,
                                         batch_size=1,
                                         img_height=img_height,
                                         img_width=img_width) 

# test_dir_1=cats_dogs/kagglecatsanddogs_3367a/PetImages/data_cat_dogs_100/test_100/'
# test_dir_2='cats_dogs/kagglecatsanddogs_3367a/PetImages/data_cat_dogs_100/test_gray_100/'

testgenerator=generate_generator_multiple(test_imgen,
                                        dir1=test_dir_1,
                                        dir2=test_dir_2,
                                        batch_size=1,
                                        img_height=img_height,
                                        img_width=img_width)      

model_ckpt = ModelCheckpoint('multi_input_model_for_ovs1_'+str(img_height)+'_.h5', monitor='val_loss', save_weights_only=False,save_best_only=True, period=1)

sgd =optimizers.SGD(lr=0.001)

model.compile(loss='categorical_crossentropy', optimizer = sgd,metrics=['accuracy'])  

model.summary()

history=model.fit_generator(inputgenerator,
                      steps_per_epoch=1000,
                      epochs = 20,
                      validation_data = testgenerator,
                      validation_steps = 500,
                      use_multiprocessing=True,
                      shuffle=False)`

after running above model architecture i am getting following logs:
1000/1000 [==============================] - 186s 186ms/step - loss: 0.0047 - acc: 1.0000 - val_loss: 2.0257e-04 - val_acc: 1.0000
Epoch 2/20
1000/1000 [==============================] - 202s 202ms/step - loss: 1.3929e-04 - acc: 1.0000 - val_loss: 9.1131e-05 - val_acc: 1.0000
Epoch 3/20
1000/1000 [==============================] - 208s 208ms/step - loss: 7.2703e-05 - acc: 1.0000 - val_loss: 5.5353e-05 - val_acc: 1.0000
Epoch 4/20
1000/1000 [==============================] - 211s 211ms/step - loss: 4.4895e-05 - acc: 1.0000 - val_loss: 3.7457e-05 - val_acc: 1.0000
Epoch 5/20
1000/1000 [==============================] - 208s 208ms/step - loss: 3.6787e-05 - acc: 1.0000 - val_loss: 6.6418 - val_acc: 0.4140
Epoch 6/20
1000/1000 [==============================] - 212s 212ms/step - loss: 0.0300 - acc: 0.9950 - val_loss: 9.9545e-04 - val_acc: 1.0000
Epoch 7/20
1000/1000 [==============================] - 206s 206ms/step - loss: 4.7121e-04 - acc: 1.0000 - val_loss: 2.0189e-04 - val_acc: 1.0000
Epoch 8/20
1000/1000 [==============================] - 213s 213ms/step - loss: 1.5082e-04 - acc: 1.0000 - val_loss: 1.0653e-04 - val_acc: 1.0000
Epoch 9/20
1000/1000 [==============================] - 199s 199ms/step - loss: 8.2482e-05 - acc: 1.0000 - val_loss: 10.5882 - val_acc: 0.0000e+00
Epoch 10/20
1000/1000 [==============================] - 204s 204ms/step - loss: 0.0328 - acc: 0.9930 - val_loss: 0.0026 - val_acc: 1.0000
Epoch 11/20
1000/1000 [==============================] - 199s 199ms/step - loss: 6.6525e-04 - acc: 1.0000 - val_loss: 2.3873e-04 - val_acc: 1.0000
Epoch 12/20
1000/1000 [==============================] - 200s 200ms/step - loss: 1.6756e-04 - acc: 1.0000 - val_loss: 1.0148e-04 - val_acc: 1.0000
Epoch 13/20
1000/1000 [==============================] - 200s 200ms/step - loss: 8.6155e-05 - acc: 1.0000 - val_loss: 6.0093 - val_acc: 0.4140
Epoch 14/20
1000/1000 [==============================] - 199s 199ms/step - loss: 5.6155e-05 - acc: 1.0000 - val_loss: 10.5919 - val_acc: 0.0000e+00
Epoch 15/20

**Comment by riyaj8888** [comment]
please help why this happened ..is there wrong in model or training

**Comment by dabasajay** [comment]
One thing I noticed that if you want to concatenate the flattened tensors 1 and 2, don't use `Add()` as it actually performs addition, not concatenation. Use `Concatenate` instead.

**Comment by riyaj8888** [comment]
thanks for replying but i already tried that "concatenation" same results were returned ....
still facing same issue... help will appreciated 

thanks

**Comment by dabasajay** [comment]
Well here are some things to try:
- Currently, your model is training with batch_size=1. Try a higher batch_size like 32 or 64 inside your data generator.
- Your model is very big in terms of parameters and if the dataset is pretty small, it's overfitting to the dataset. Try a smaller model (Maybe reduce last dense layers and number of units)

**Comment by riyaj8888** [comment]
i tried by changing batch size and also reducing fc layers units but  still facing same issue . from 1st epoch it shows 
134/1000 [===>..........................] - ETA: 4:47 - loss: 0.0180 - acc: 1.0000

**Comment by dabasajay** [comment]
Just noticed that you're probably doing binary classification with labels as either a `dog` or a `cat` right? If yes, do the following changes and let me know how it goes:
- change number of units in the output layer=1 and change activation to `sigmoid`
    - `output = Dense(1, activation = 'sigmoid')(dense3)`
- change loss to `binary_crossentropy`
    - `model.compile(loss='binary_crossentropy', optimizer = sgd,metrics=['accuracy'])`

**Comment by riyaj8888** [comment]
Doesn't make any sense... What do u think what's reason of this behaviour

**Comment by dabasajay** [comment]
Well, I've seen someone encountered a problem like this and the changes I suggested in my last comment worked for her so I'm just saying just try and if it works, we'll discuss further otherwise move on to next guess :stuck_out_tongue:

**Comment by riyaj8888** [comment]
Not worked .... same issue

**Comment by dabasajay** [comment]
Ok last try, can I see your new code for `generate_generator_multiple` where you used batch_size is >1?

**Comment by riyaj8888** [comment]
Will u share ur email id so I can send u my whole code and please if
possible go through it and help if u find anything suspicious

**Comment by dabasajay** [comment]
Sure, it's se.dabasajay@gmail.com

**Comment by riyaj8888** [comment]
sent the code please look at it

thanks
